osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyMainCirclePiece.cs

```diff
@@ -16,63 +16,61 @@
 using osuTK;
 using osuTK.Graphics;
 
+#nullable enable
+
 namespace osu.Game.Rulesets.Osu.Skinning.Legacy
 {
     public class LegacyMainCirclePiece : CompositeDrawable
     {
         public override bool RemoveCompletedTransforms => false;
 
-        private readonly string priorityLookup;
-        private readonly bool hasNumber;
-
-        public LegacyMainCirclePiece(string priorityLookup = null, bool hasNumber = true)
-        {
-            this.priorityLookup = priorityLookup;
-            this.hasNumber = hasNumber;
+        /// <summary>
+        /// A prioritised prefix to perform texture lookups with.
+        /// </summary>
+        private readonly string? priorityLookupPrefix;
 
-            Size = new Vector2(OsuHitObject.OBJECT_RADIUS * 2);
-        }
+        private readonly bool hasNumber;
 
-        private Drawable hitCircleSprite;
+        private Drawable hitCircleSprite = null!;
 
-        protected Container OverlayLayer { get; private set; }
+        protected Container OverlayLayer { get; private set; } = null!;
 
-        private Drawable hitCircleOverlay;
-        private SkinnableSpriteText hitCircleText;
+        private Drawable hitCircleOverlay = null!;
+        private SkinnableSpriteText hitCircleText = null!;
 
         private readonly Bindable<Color4> accentColour = new Bindable<Color4>();
         private readonly IBindable<int> indexInCurrentCombo = new Bindable<int>();
 
-        [Resolved]
-        private DrawableHitObject drawableObject { get; set; }
+        [Resolved(canBeNull: true)]
+        private DrawableHitObject? drawableObject { get; set; }
```

Comment:
Where's this `canBeNull: true` spec coming from? The dependency was not previously allowed to be null.

Is it because of the pre-existing null check in `Dispose()` triggering a "never null" inspection or something?

Reply:
More like because it’s used in a test scene in isolation, therefore needs to not require the DHO as a dependency.

Reply:
I would have gone the other way and cached one for the sake of it but I'm not sure I care enough to argue this further? Like it seems weird to me that integral dependencies are being marked nullable just for the sake of a test, but alas.

Reply:
Sure, probably better if feasible.

Reply:
After attempting to bring this, I didn't take into account that this would actually cause the aforementioned inspection error ("pre-existing null check in Dispose() triggering a "never null" inspection"):

![](https://cdn.discordapp.com/attachments/188630652340404224/966081850571366410/unknown.png)

Can still push with an inspection disable comment and a TODO:
```csharp
         protected override void Dispose(bool isDisposing)
         {
             base.Dispose(isDisposing);
 
             // ReSharper disable once ConditionIsAlwaysTrueOrFalseAccordingToNullableAPIContract
             // todo: we probably want to handle this in a better way somehow without disabling inspection.
             if (drawableObject != null)
                 drawableObject.ApplyCustomUpdateState -= updateStateTransforms;
         }
```
but not 100% sure if we just want to leave it like that.